File: android/sdk/src/main/kotlin/com/ary/push/internal/notification/NotificationParser.kt
package com.ary.push.internal.notification

import com.ary.push.model.PushNotification
import com.google.firebase.messaging.RemoteMessage

/**
 * Maps an FCM [RemoteMessage] onto the platform-neutral [PushNotification].
 *
 * FCM messages arrive in three shapes and the SDK has to normalise all of them:
 *
 *  * **Notification messages** carry a `notification` block. When the app is backgrounded the
 *    system renders these itself and the app is only told on tap.
 *  * **Data messages** carry only `data`, always reach the service, and must be rendered by the
 *    SDK.
 *  * **Mixed messages** carry both.
 *
 * Host applications should prefer data-only messages: they are the only shape whose foreground,
 * background and terminated behaviour the SDK can make identical. See
 * docs/NOTIFICATION_LIFECYCLE.md.
 */
internal object NotificationParser {

    /** Data keys understood as a fallback when there is no `notification` block. */
    private const val KEY_TITLE = "title"
    private const val KEY_BODY = "body"
    private const val KEY_IMAGE = "image"
    private const val KEY_IMAGE_URL = "image_url"
    private const val KEY_CHANNEL_ID = "channel_id"
    private const val KEY_NOTIFICATION_ID = "notification_id"
    private const val KEY_ID = "id"

    fun parse(message: RemoteMessage, wasForeground: Boolean): PushNotification {
        val data = message.data
        val notification = message.notification

        return PushNotification(
            id = resolveId(message),
            title = notification?.title ?: data[KEY_TITLE],
            body = notification?.body ?: data[KEY_BODY],
            imageUrl = notification?.imageUrl?.toString()
                ?: data[KEY_IMAGE_URL]
                ?: data[KEY_IMAGE],
            data = data.toMap(),
            receivedAt = System.currentTimeMillis(),
            sentAt = message.sentTime.takeIf { it > 0 },
            channelId = notification?.channelId ?: data[KEY_CHANNEL_ID],
            collapseKey = message.collapseKey,
            actionId = null,
            wasForeground = wasForeground
        )
    }

    /**
     * A stable identity for deduplication.
     *
     * Preference order matters. A sender-supplied id groups a logical message across resends;
     * FCM's message id is unique per delivery attempt; the content hash is the last resort for
     * senders that supply neither, and is what makes deduplication work at all for them.
     */
    fun resolveId(message: RemoteMessage): String {
        message.data[KEY_NOTIFICATION_ID]?.takeIf { it.isNotBlank() }?.let { return it }
        message.data[KEY_ID]?.takeIf { it.isNotBlank() }?.let { return it }
        message.messageId?.takeIf { it.isNotBlank() }?.let { return it }
        return contentHash(message)
    }

    private fun contentHash(message: RemoteMessage): String {
        val fingerprint = buildString {
            append(message.notification?.title.orEmpty()).append('\u0000')
            append(message.notification?.body.orEmpty()).append('\u0000')
            append(message.collapseKey.orEmpty()).append('\u0000')
            append(message.sentTime).append('\u0000')
            message.data.toSortedMap().forEach { (key, value) ->
                append(key).append('=').append(value).append(';')
            }
        }
        return "hash-${fingerprint.hashCode()}"
    }

    /** Android notification ids are ints; this derives a stable one from the message id. */
    fun systemNotificationId(notification: PushNotification): Int {
        // Collapse key first: two messages sharing one deliberately replace each other in the
        // shade, which is what a collapse key is for.
        val basis = notification.collapseKey?.takeIf { it.isNotBlank() } ?: notification.id
        return basis.hashCode()
    }
}

File: ios/Sources/ARYPush/Internal/NotificationParser.swift
import Foundation
import UserNotifications

/// Maps an APNs payload onto the platform-neutral ``PushNotification``.
///
/// The same payload reaches the SDK through three different callbacks, and this is where they
/// are made to look identical:
///
/// * `userNotificationCenter(_:willPresent:)` for a foreground notification;
/// * `userNotificationCenter(_:didReceive:)` for a tap, including from a terminated launch;
/// * `application(_:didReceiveRemoteNotification:)` for a silent, content-available message.
enum NotificationParser {

    private static let apsKey = "aps"
    private static let notificationIdKeys = ["notification_id", "id", "gcm.message_id"]

    /// Payload keys added by Google's SDKs that are transport plumbing, not application data.
    private static let reservedPrefixes = ["gcm.", "google.", "aps"]

    static func parse(
        userInfo: [AnyHashable: Any],
        wasForeground: Bool,
        actionId: String? = nil
    ) -> PushNotification {
        let aps = userInfo[apsKey] as? [String: Any] ?? [:]
        let alert = aps["alert"]

        var title: String?
        var body: String?
        if let alertString = alert as? String {
            body = alertString
        } else if let alertDict = alert as? [String: Any] {
            title = alertDict["title"] as? String
            body = (alertDict["body"] as? String) ?? (alertDict["subtitle"] as? String)
        }

        let data = customData(from: userInfo)

        return PushNotification(
            id: resolveId(userInfo: userInfo, data: data),
            title: title ?? data["title"],
            body: body ?? data["body"],
            imageURL: imageURL(userInfo: userInfo, data: data),
            data: data,
            receivedAt: Date(),
            threadIdentifier: aps["thread-id"] as? String,
            categoryIdentifier: aps["category"] as? String,
            actionId: actionId,
            wasForeground: wasForeground
        )
    }

    /// Maps a delivered `UNNotification`, preserving the identifier iOS assigned it.
    static func parse(
        notification: UNNotification,
        wasForeground: Bool,
        actionId: String? = nil
    ) -> PushNotification {
        let parsed = parse(
            userInfo: notification.request.content.userInfo,
            wasForeground: wasForeground,
            actionId: actionId
        )
        let content = notification.request.content
        return PushNotification(
            id: parsed.id,
            title: parsed.title ?? content.title.nonEmpty,
            body: parsed.body ?? content.body.nonEmpty,
            imageURL: parsed.imageURL,
            data: parsed.data,
            receivedAt: notification.date,
            threadIdentifier: parsed.threadIdentifier ?? content.threadIdentifier.nonEmpty,
            categoryIdentifier: parsed.categoryIdentifier ?? content.categoryIdentifier.nonEmpty,
            actionId: actionId,
            wasForeground: wasForeground
        )
    }

    /// The application-visible payload, with transport plumbing removed.
    static func customData(from userInfo: [AnyHashable: Any]) -> [String: String] {
        var result: [String: String] = [:]
        for (rawKey, value) in userInfo {
            guard let key = rawKey as? String else { continue }
            guard !reservedPrefixes.contains(where: { key == $0 || key.hasPrefix($0) }) else {
                continue
            }
            if let string = value as? String {
                result[key] = string
            } else if let data = try? JSONSerialization.data(
                withJSONObject: JSONSanitizer.sanitize(value)
            ), let encoded = String(data: data, encoding: .utf8) {
                // Nested objects are handed over as JSON rather than dropped, so a host
                // application can still read them.
                result[key] = encoded
            } else {
                result[key] = String(describing: value)
            }
        }
        return result
    }

    /// A stable identity for deduplication.
    ///
    /// Preference order matters. A sender-supplied id groups a logical message across resends;
    /// the transport's own id is unique per delivery attempt; the content hash is the last
    /// resort for senders that supply neither, and is what makes deduplication work for them.
    static func resolveId(userInfo: [AnyHashable: Any], data: [String: String]) -> String {
        for key in notificationIdKeys {
            if let value = (userInfo[key] as? String) ?? data[key], !value.isEmpty {
                return value
            }
        }
        return contentHash(userInfo: userInfo)
    }

    private static func imageURL(
        userInfo: [AnyHashable: Any],
        data: [String: String]
    ) -> String? {
        if let options = userInfo["fcm_options"] as? [String: Any],
           let image = options["image"] as? String {
            return image
        }
        return data["image_url"] ?? data["image"]
    }

    private static func contentHash(userInfo: [AnyHashable: Any]) -> String {
        let fingerprint = userInfo
            .compactMap { key, value -> String? in
                guard let key = key as? String else { return nil }
                return "\(key)=\(String(describing: value))"
            }
            .sorted()
            .joined(separator: ";")
        return "hash-\(StableHash.digest(fingerprint))"
    }
}

private extension String {
    var nonEmpty: String? { isEmpty ? nil : self }
}
